x-pack/plugins/cases/public/components/files/file_type.tsx

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import React from 'react';
+
+import type { FileJSON } from '@kbn/shared-ux-file-types';
+
+import type {
+  ExternalReferenceAttachmentType,
+  ExternalReferenceAttachmentViewProps,
+} from '../../client/attachment_framework/types';
+
+import { AttachmentActionType } from '../../client/attachment_framework/types';
+import { FILE_ATTACHMENT_TYPE } from '../../../common/api';
+import { FileDownloadButtonIcon } from './file_download_button_icon';
+import { FileNameLink } from './file_name_link';
+import { FilePreview } from './file_preview';
+import * as i18n from './translations';
+import { isImage, isValidFileExternalReferenceMetadata } from './utils';
+import { useFilePreview } from './use_file_preview';
+import { FileDeleteButtonIcon } from './file_delete_button_icon';
+
+interface FileAttachmentEventProps {
+  file: FileJSON;
+}
+
+const FileAttachmentEvent = ({ file }: FileAttachmentEventProps) => {
+  const { isPreviewVisible, showPreview, closePreview } = useFilePreview();
+
+  return (
+    <>
+      {i18n.ADDED}
+      <FileNameLink file={file} showPreview={showPreview} />
+      {isPreviewVisible && <FilePreview closePreview={closePreview} selectedFile={file} />}
+    </>
+  );
+};
+
+FileAttachmentEvent.displayName = 'FileAttachmentEvent';
+
+function getFileDownloadButtonIcon(fileId: string) {
+  return <FileDownloadButtonIcon fileId={fileId} />;
+}
+
+function getFileDeleteButtonIcon(caseId: string, fileId: string) {
+  return <FileDeleteButtonIcon caseId={caseId} fileId={fileId} />;
+}
+
+const getFileAttachmentActions = ({ caseId, fileId }: { caseId: string; fileId: string }) => [
+  {
+    type: AttachmentActionType.CUSTOM as const,
+    isPrimary: true,
+    render: () => getFileDownloadButtonIcon(fileId),
+    label: i18n.DOWNLOAD_FILE,
+  },
+  {
+    type: AttachmentActionType.CUSTOM as const,
+    isPrimary: true,
+    render: () => getFileDeleteButtonIcon(caseId, fileId),
+    label: i18n.DELETE_FILE,
+  },
+];
+
+const getFileAttachmentViewObject = (props: ExternalReferenceAttachmentViewProps) => {
+  if (!isValidFileExternalReferenceMetadata(props.externalReferenceMetadata)) {
+    return {
+      type: 'regular',
+      event: i18n.ADDED_UNKNOWN_FILE,
+      timelineAvatar: 'document',
+      hideDefaultActions: true,
+    };
+  }
+
+  const fileId = props.externalReferenceId;
+  const caseId = props.caseData.id;
+
+  // @ts-ignore
+  const partialFileJSON = props.externalReferenceMetadata?.files[0] as Partial<FileJSON>;
+
+  const file = {
+    id: fileId,
+    ...partialFileJSON,
+  } as FileJSON;
```

Comment:
I know I mentioned this on a previous PR, sorry I didn't follow up on your response. I think it could be dangerous to cast this back to a FileJSON and pass that around. My concern is the components are expecting a legitimate FileJSON object and the fields in `externalReferenceMetadata` don't include at least one of the fields (`alt` I think).

I think we can avoid the `ts-ignore` and the cast here with the following changes

First the `isValidFileExternalReferenceMetadata` call needs to be a typeguard instead of only returning a boolean.

```
export const isValidFileExternalReferenceMetadata = (
  externalReferenceMetadata: CommentRequestExternalReferenceType['externalReferenceMetadata']
): externalReferenceMetadata is FileAttachmentMetadata => {
  return (
    FileAttachmentMetadataRt.is(externalReferenceMetadata) &&
    externalReferenceMetadata?.files?.length >= 1
  );
};
```

Then we can create our own type say `DownloadableFile` that includes everything from the metadata we store and the `id`.

```
  // a better option instead of this wild ['files'][0] nonsense haha would be to extract the io-ts type and use that within FileAttachmentMetadataRt and also reference it here
  type DownloadableFile = FileAttachmentMetadata['files'][0] & { id: string };

  const fileMetadata = props.externalReferenceMetadata.files[0];

  const file: DownloadableFile = {
    id: fileId,
    ...fileMetadata,
  };
```

Now when we pass `file` around we can be more confident of what it actually contains. The downside here is that we'll need update the types for all the child components.

Reply:
Good input, I did it 👍

> The downside here is that we'll need update the types for all the child components.

I updated the types of child components to be more restrictive.

In some places, I was requiring `FileJSON` where we just needed like `{ id: string; name: string }`.

I think it is better like this and since these fields are common between `DownloadableFile` and `FileJSON` it will work in both places where they are used.

Sometimes we render these components using `metadata: DownloadableFile` and sometimes after `apiCallResponse: FileJSON`.